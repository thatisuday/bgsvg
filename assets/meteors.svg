<svg xmlns="http://www.w3.org/2000/svg" width="800" height="300" viewBox="0 0 800 300">
    <defs>
        <linearGradient id="backgroundGradient" x1="0%" y1="0%" x2="100%" y2="100%">
            <stop offset="0%" style="stop-color: #000000; stop-opacity: 1"></stop>
            <stop offset="100%" style="stop-color: #13137f; stop-opacity: 1"></stop>
        </linearGradient>
        <linearGradient id="gradient">
            <stop offset="0%" stop-color="rgba(186,128,3,0)"></stop>
            <stop offset="100%" stop-color="rgb(186,128,3)"></stop>
        </linearGradient>
        <linearGradient id="gradient_rev">
            <stop offset="0%" stop-color="rgb(186,128,3)"></stop>
            <stop offset="100%" stop-color="rgba(186,128,3,0)"></stop>
        </linearGradient>
    </defs>
    <rect width="800" height="300" fill="url(#backgroundGradient)"></rect>
    <line x1="-2" y1="48" x2="187" y2="237" stroke-width="4" stroke-linecap="round"
        stroke="url(#gradient_rev)"></line>
    <line x1="-10" y1="237" x2="176" y2="423" stroke-width="4" stroke-linecap="round"
        stroke="url(#gradient)"></line>
    <line x1="34" y1="65" x2="193" y2="224" stroke-width="5" stroke-linecap="round"
        stroke="url(#gradient)"></line>
    <line x1="39" y1="250" x2="225" y2="436" stroke-width="4" stroke-linecap="round"
        stroke="url(#gradient)"></line>
    <line x1="70" y1="-61" x2="234" y2="103" stroke-width="5" stroke-linecap="round"
        stroke="url(#gradient)"></line>
    <line x1="84" y1="316" x2="241" y2="473" stroke-width="5" stroke-linecap="round"
        stroke="url(#gradient)"></line>
    <line x1="126" y1="-21" x2="301" y2="154" stroke-width="4" stroke-linecap="round"
        stroke="url(#gradient)"></line>
    <line x1="121" y1="235" x2="273" y2="387" stroke-width="4" stroke-linecap="round"
        stroke="url(#gradient_rev)"></line>
    <line x1="164" y1="-7" x2="322" y2="151" stroke-width="4" stroke-linecap="round"
        stroke="url(#gradient)"></line>
    <line x1="156" y1="372" x2="317" y2="533" stroke-width="4" stroke-linecap="round"
        stroke="url(#gradient)"></line>
    <line x1="199" y1="-57" x2="377" y2="121" stroke-width="4" stroke-linecap="round"
        stroke="url(#gradient)"></line>
    <line x1="191" y1="338" x2="345" y2="492" stroke-width="5" stroke-linecap="round"
        stroke="url(#gradient)"></line>
    <line x1="234" y1="-55" x2="416" y2="127" stroke-width="4" stroke-linecap="round"
        stroke="url(#gradient)"></line>
    <line x1="240" y1="290" x2="435" y2="485" stroke-width="6" stroke-linecap="round"
        stroke="url(#gradient_rev)"></line>
    <line x1="275" y1="74" x2="440" y2="239" stroke-width="4" stroke-linecap="round"
        stroke="url(#gradient)"></line>
    <line x1="278" y1="319" x2="458" y2="499" stroke-width="4" stroke-linecap="round"
        stroke="url(#gradient_rev)"></line>
    <line x1="314" y1="-35" x2="509" y2="160" stroke-width="4" stroke-linecap="round"
        stroke="url(#gradient)"></line>
    <line x1="314" y1="352" x2="507" y2="545" stroke-width="5" stroke-linecap="round"
        stroke="url(#gradient)"></line>
    <line x1="362" y1="31" x2="517" y2="186" stroke-width="5" stroke-linecap="round"
        stroke="url(#gradient_rev)"></line>
    <line x1="357" y1="375" x2="543" y2="561" stroke-width="4" stroke-linecap="round"
        stroke="url(#gradient)"></line>
    <line x1="396" y1="5" x2="570" y2="179" stroke-width="5" stroke-linecap="round"
        stroke="url(#gradient)"></line>
    <line x1="404" y1="304" x2="595" y2="495" stroke-width="4" stroke-linecap="round"
        stroke="url(#gradient)"></line>
    <line x1="432" y1="2" x2="589" y2="159" stroke-width="4" stroke-linecap="round"
        stroke="url(#gradient_rev)"></line>
    <line x1="446" y1="260" x2="643" y2="457" stroke-width="3" stroke-linecap="round"
        stroke="url(#gradient)"></line>
    <line x1="472" y1="-62" x2="655" y2="121" stroke-width="4" stroke-linecap="round"
        stroke="url(#gradient)"></line>
    <line x1="479" y1="307" x2="677" y2="505" stroke-width="4" stroke-linecap="round"
        stroke="url(#gradient)"></line>
    <line x1="524" y1="11" x2="683" y2="170" stroke-width="5" stroke-linecap="round"
        stroke="url(#gradient_rev)"></line>
    <line x1="517" y1="356" x2="701" y2="540" stroke-width="5" stroke-linecap="round"
        stroke="url(#gradient_rev)"></line>
    <line x1="564" y1="42" x2="745" y2="223" stroke-width="3" stroke-linecap="round"
        stroke="url(#gradient_rev)"></line>
    <line x1="555" y1="331" x2="718" y2="494" stroke-width="5" stroke-linecap="round"
        stroke="url(#gradient)"></line>
    <line x1="601" y1="17" x2="790" y2="206" stroke-width="4" stroke-linecap="round"
        stroke="url(#gradient_rev)"></line>
    <line x1="608" y1="298" x2="780" y2="470" stroke-width="5" stroke-linecap="round"
        stroke="url(#gradient)"></line>
    <line x1="631" y1="-31" x2="831" y2="169" stroke-width="5" stroke-linecap="round"
        stroke="url(#gradient_rev)"></line>
    <line x1="645" y1="300" x2="829" y2="484" stroke-width="4" stroke-linecap="round"
        stroke="url(#gradient_rev)"></line>
    <line x1="678" y1="-63" x2="842" y2="101" stroke-width="4" stroke-linecap="round"
        stroke="url(#gradient)"></line>
    <line x1="681" y1="344" x2="873" y2="536" stroke-width="4" stroke-linecap="round"
        stroke="url(#gradient_rev)"></line>
    <line x1="719" y1="39" x2="916" y2="236" stroke-width="4" stroke-linecap="round"
        stroke="url(#gradient)"></line>
    <line x1="730" y1="289" x2="924" y2="483" stroke-width="5" stroke-linecap="round"
        stroke="url(#gradient)"></line>
    <line x1="765" y1="-60" x2="951" y2="126" stroke-width="4" stroke-linecap="round"
        stroke="url(#gradient_rev)"></line>
    <line x1="760" y1="355" x2="937" y2="532" stroke-width="5" stroke-linecap="round"
        stroke="url(#gradient_rev)"></line>
    <line x1="794" y1="-51" x2="965" y2="120" stroke-width="5" stroke-linecap="round"
        stroke="url(#gradient_rev)"></line>
    <line x1="796" y1="225" x2="986" y2="415" stroke-width="5" stroke-linecap="round"
        stroke="url(#gradient_rev)"></line>
</svg>