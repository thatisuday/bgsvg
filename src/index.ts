export { meteors } from "./meteors";
export { rain } from "./rain";
