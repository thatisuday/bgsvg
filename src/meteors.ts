import { Window, type SVGSVGElement } from "happy-dom";
import { CanvasBackground, getCanvas } from "./lib/canvas";
import { SVG_NS } from "./lib/const";
import { randomNumber, randomElement, get2dPoints } from "./lib/random";
import chroma from "chroma-js";
import { createSvgGradient } from "./lib/svg";
import type { OutputType, OutputTypeSvg } from "./lib/types";
import { getOutput } from "./lib/output";

/**
 * Add a meteor to the SVG.
 */
const addMeteor = ({
    svg,
    start,
    end,
    thickness,
    gradientId,
}: {
    svg: SVGSVGElement;
    start: [number, number];
    end: [number, number];
    thickness: number;
    gradientId: string;
}) => {
    const { document } = new Window();

    const meteor = document.createElementNS(SVG_NS, "line");
    const [x1, y1] = start;
    const [x2, y2] = end;

    meteor.setAttribute("x1", x1.toString());
    meteor.setAttribute("y1", y1.toString());
    meteor.setAttribute("x2", x2.toString());
    meteor.setAttribute("y2", y2.toString());
    meteor.setAttribute("stroke-width", thickness.toString());
    meteor.setAttribute("stroke-linecap", "round");

    meteor.setAttribute("stroke", `url(#${gradientId})`);
    svg.appendChild(meteor);
};

export async function meteors(options: {
    width: number;
    height: number;
    background: CanvasBackground;
    color: string;
    min?: number;
    max?: number;
    thickness?: number;
    bidirectional?: boolean;
    output?: OutputTypeSvg;
}): Promise<string>;

export async function meteors(options: {
    width: number;
    height: number;
    background: CanvasBackground;
    color: string;
    min?: number;
    max?: number;
    thickness?: number;
    bidirectional?: boolean;
    output: Exclude<OutputType, OutputTypeSvg>;
}): Promise<Buffer>;

/**
 * Generate an SVG with meteors.
 */
export async function meteors({
    width,
    height,
    background,
    color,
    densityX = 20,
    densityY = 1,
    thickness = 4,
    bidirectional = true,
    output = {
        type: "svg",
    },
}: {
    width: number;
    height: number;
    background: CanvasBackground;
    color: string;
    densityX?: number;
    densityY?: number;
    thickness?: number;
    bidirectional?: boolean;
    output?: OutputType;
}): Promise<string | Buffer> {
    const { document } = new Window();
    const svg = getCanvas({ width, height, background });
    const defs = svg.getElementsByTagName("defs")?.[0];

    const gradient = createSvgGradient({
        id: "gradient",
        startColor: chroma(color).alpha(0).css(),
        endColor: chroma(color).alpha(1).css(),
    });

    if (!defs) {
        const defs = document.createElementNS(SVG_NS, "defs");
        defs.appendChild(gradient);
        svg.appendChild(defs);
    } else {
        defs.appendChild(gradient);
    }

    // add reverse gradient
    if (bidirectional) {
        const gradientRev = createSvgGradient({
            id: "gradient_rev",
            startColor: chroma(color).alpha(1).css(),
            endColor: chroma(color).alpha(0).css(),
        });

        if (!defs) {
            const defs = document.createElementNS(SVG_NS, "defs");
            defs.appendChild(gradientRev);
            svg.appendChild(defs);
        } else {
            defs.appendChild(gradientRev);
        }
    }

    // generate coordinate points for drops
    const points = get2dPoints({
        width,
        height,
        densityX,
        densityY,
        varianceFactor: 4,
    });

    for (const [x, y] of points) {
        const x1 = x;
        const y1 = y;
        const length = Math.round(randomNumber(height / densityY / 2, (2 * height) / densityY / 3));
        const x2 = x1 + length;
        const y2 = y1 + length;

        const meteorGradientId = bidirectional ? randomElement("gradient", "gradient_rev") : "gradient";

        addMeteor({
            svg,
            start: [x1, y1],
            end: [x2, y2],
            thickness: Math.round(randomNumber(thickness, thickness + (Math.random() - 0.25) * thickness)),
            gradientId: meteorGradientId,
        });
    }

    return getOutput({
        width,
        height,
        svg,
        output,
    });
}
