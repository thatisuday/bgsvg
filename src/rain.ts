import { Window, type SVGSVGElement } from "happy-dom";
import { CanvasBackground, getCanvas } from "./lib/canvas";
import { SVG_NS } from "./lib/const";
import { randomNumber, get2dPoints } from "./lib/random";
import chroma from "chroma-js";
import { createSvgGradient } from "./lib/svg";
import type { OutputType, OutputTypeSvg } from "./lib/types";
import { getOutput } from "./lib/output";

/**
 * Add a rain drop to the SVG.
 */
const addDrop = ({
    svg,
    start,
    end,
    thickness,
    gradientId,
}: {
    svg: SVGSVGElement;
    start: [number, number];
    end: [number, number];
    thickness: number;
    gradientId: string;
}) => {
    const { document } = new Window();

    const meteor = document.createElementNS(SVG_NS, "line");
    const [x1, y1] = start;
    const [x2, y2] = end;

    meteor.setAttribute("x1", x1.toString());
    meteor.setAttribute("y1", y1.toString());
    meteor.setAttribute("x2", x2.toString());
    meteor.setAttribute("y2", y2.toString());
    meteor.setAttribute("stroke-width", thickness.toString());
    meteor.setAttribute("stroke-linecap", "round");

    meteor.setAttribute("stroke", `url(#${gradientId})`);
    svg.appendChild(meteor);
};

export async function rain(options: {
    width: number;
    height: number;
    background: CanvasBackground;
    color: string;
    dropsX?: number;
    dropsY?: number;
    thickness?: number;
    output?: OutputTypeSvg;
}): Promise<string>;

export async function rain(options: {
    width: number;
    height: number;
    background: CanvasBackground;
    color: string;
    dropsX?: number;
    dropsY?: number;
    thickness?: number;
    output: Exclude<OutputType, OutputTypeSvg>;
}): Promise<Buffer>;

/**
 * Generate an SVG with rain.
 */
export async function rain({
    width,
    height,
    background,
    color,
    densityX = 20,
    densityY = 15,
    thickness = 1,
    randomness = 4,
    output = {
        type: "svg",
    },
}: {
    width: number;
    height: number;
    background: CanvasBackground;
    color: string;
    densityX?: number;
    densityY?: number;
    thickness?: number;
    randomness?: number;
    output?: OutputType;
}): Promise<string | Buffer> {
    const { document } = new Window();
    const svg = getCanvas({ width, height, background });

    // generate coordinate points for drops
    const points = get2dPoints({
        width,
        height,
        densityX,
        densityY,
        varianceFactor: randomness,
    });

    for (const [x, y] of points) {
        const x1 = x;
        const y1 = y;
        const length = Math.round(randomNumber(height / densityY / 2, (2 * height) / densityY / 3));
        const x2 = x1 + length;
        const y2 = y1 + length;
        const gradientId = `gradient_${randomNumber(1, 10000000)}`;

        const gradient = createSvgGradient({
            id: gradientId,
            startColor: chroma(color).alpha(randomNumber(0, 0.25)).css(),
            endColor: chroma(color).alpha(randomNumber(0.25, 1)).css(),
        });

        const defs = svg.getElementsByTagName("defs")?.[0];
        if (!defs) {
            const defs = document.createElementNS(SVG_NS, "defs");
            defs.appendChild(gradient);
            svg.appendChild(defs);
        } else {
            defs.appendChild(gradient);
        }

        addDrop({
            svg,
            start: [x1, y1],
            end: [x2, y2],
            thickness: randomNumber(thickness, thickness + (Math.random() - 0.25) * thickness),
            gradientId,
        });
    }

    return getOutput({
        width,
        height,
        svg,
        output,
    });
}
