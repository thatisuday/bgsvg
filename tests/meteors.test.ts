import { meteors } from "../src/meteors";

describe("meteors", () => {
    it("should match the snapshot", async () => {
        const svg = await meteors({
            width: 500,
            height: 200,
            background: "#000",
            color: "#fff",
        });

        expect(svg).toMatchSnapshot();
    });
});
