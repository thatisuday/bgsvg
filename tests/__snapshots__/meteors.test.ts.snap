// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`meteors should match the snapshot 1`] = `"<svg xmlns="http://www.w3.org/2000/svg" width="500" height="200" viewbox="0 0 500 200"><rect width="500" height="200" fill="#000"></rect><defs><linearGradient id="gradient"><stop offset="0%" stop-color="rgba(255,255,255,0)"></stop><stop offset="100%" stop-color="rgb(255,255,255)"></stop></linearGradient></defs><defs><linearGradient id="gradient_rev"><stop offset="0%" stop-color="rgb(255,255,255)"></stop><stop offset="100%" stop-color="rgba(255,255,255,0)"></stop></linearGradient></defs><line x1="-6" y1="-4" x2="118" y2="120" stroke-width="4" stroke-linecap="round" stroke="url(#gradient_rev)"></line><line x1="-6" y1="204" x2="98" y2="308" stroke-width="5" stroke-linecap="round" stroke="url(#gradient_rev)"></line><line x1="29" y1="-33" x2="138" y2="76" stroke-width="4" stroke-linecap="round" stroke="url(#gradient)"></line><line x1="25" y1="224" x2="150" y2="349" stroke-width="4" stroke-linecap="round" stroke="url(#gradient)"></line><line x1="52" y1="-3" x2="169" y2="114" stroke-width="5" stroke-linecap="round" stroke="url(#gradient)"></line><line x1="49" y1="183" x2="161" y2="295" stroke-width="5" stroke-linecap="round" stroke="url(#gradient_rev)"></line><line x1="81" y1="48" x2="195" y2="162" stroke-width="7" stroke-linecap="round" stroke="url(#gradient_rev)"></line><line x1="80" y1="168" x2="208" y2="296" stroke-width="4" stroke-linecap="round" stroke="url(#gradient_rev)"></line><line x1="99" y1="23" x2="225" y2="149" stroke-width="4" stroke-linecap="round" stroke="url(#gradient_rev)"></line><line x1="104" y1="158" x2="216" y2="270" stroke-width="5" stroke-linecap="round" stroke="url(#gradient_rev)"></line><line x1="129" y1="-28" x2="252" y2="95" stroke-width="5" stroke-linecap="round" stroke="url(#gradient)"></line><line x1="130" y1="182" x2="259" y2="311" stroke-width="5" stroke-linecap="round" stroke="url(#gradient)"></line><line x1="149" y1="-39" x2="275" y2="87" stroke-width="5" stroke-linecap="round" stroke="url(#gradient)"></line><line x1="145" y1="165" x2="258" y2="278" stroke-width="4" stroke-linecap="round" stroke="url(#gradient)"></line><line x1="180" y1="5" x2="305" y2="130" stroke-width="5" stroke-linecap="round" stroke="url(#gradient)"></line><line x1="173" y1="215" x2="293" y2="335" stroke-width="4" stroke-linecap="round" stroke="url(#gradient)"></line><line x1="204" y1="-45" x2="306" y2="57" stroke-width="6" stroke-linecap="round" stroke="url(#gradient)"></line><line x1="194" y1="234" x2="316" y2="356" stroke-width="4" stroke-linecap="round" stroke="url(#gradient)"></line><line x1="223" y1="-9" x2="340" y2="108" stroke-width="3" stroke-linecap="round" stroke="url(#gradient_rev)"></line><line x1="220" y1="222" x2="335" y2="337" stroke-width="4" stroke-linecap="round" stroke="url(#gradient)"></line><line x1="250" y1="16" x2="367" y2="133" stroke-width="3" stroke-linecap="round" stroke="url(#gradient_rev)"></line><line x1="246" y1="184" x2="366" y2="304" stroke-width="4" stroke-linecap="round" stroke="url(#gradient_rev)"></line><line x1="274" y1="40" x2="402" y2="168" stroke-width="4" stroke-linecap="round" stroke="url(#gradient)"></line><line x1="272" y1="168" x2="405" y2="301" stroke-width="4" stroke-linecap="round" stroke="url(#gradient_rev)"></line><line x1="305" y1="-7" x2="411" y2="99" stroke-width="4" stroke-linecap="round" stroke="url(#gradient)"></line><line x1="303" y1="195" x2="415" y2="307" stroke-width="4" stroke-linecap="round" stroke="url(#gradient)"></line><line x1="321" y1="14" x2="451" y2="144" stroke-width="4" stroke-linecap="round" stroke="url(#gradient)"></line><line x1="326" y1="202" x2="445" y2="321" stroke-width="4" stroke-linecap="round" stroke="url(#gradient)"></line><line x1="355" y1="-13" x2="473" y2="105" stroke-width="4" stroke-linecap="round" stroke="url(#gradient_rev)"></line><line x1="353" y1="183" x2="486" y2="316" stroke-width="3" stroke-linecap="round" stroke="url(#gradient_rev)"></line><line x1="370" y1="-31" x2="487" y2="86" stroke-width="4" stroke-linecap="round" stroke="url(#gradient_rev)"></line><line x1="373" y1="156" x2="479" y2="262" stroke-width="4" stroke-linecap="round" stroke="url(#gradient_rev)"></line><line x1="400" y1="15" x2="519" y2="134" stroke-width="5" stroke-linecap="round" stroke="url(#gradient)"></line><line x1="403" y1="202" x2="525" y2="324" stroke-width="6" stroke-linecap="round" stroke="url(#gradient_rev)"></line><line x1="431" y1="-17" x2="539" y2="91" stroke-width="4" stroke-linecap="round" stroke="url(#gradient)"></line><line x1="421" y1="245" x2="537" y2="361" stroke-width="4" stroke-linecap="round" stroke="url(#gradient)"></line><line x1="453" y1="41" x2="568" y2="156" stroke-width="4" stroke-linecap="round" stroke="url(#gradient)"></line><line x1="451" y1="246" x2="552" y2="347" stroke-width="5" stroke-linecap="round" stroke="url(#gradient_rev)"></line><line x1="473" y1="30" x2="593" y2="150" stroke-width="4" stroke-linecap="round" stroke="url(#gradient_rev)"></line><line x1="480" y1="197" x2="600" y2="317" stroke-width="4" stroke-linecap="round" stroke="url(#gradient)"></line><line x1="505" y1="-23" x2="620" y2="92" stroke-width="6" stroke-linecap="round" stroke="url(#gradient_rev)"></line><line x1="495" y1="236" x2="607" y2="348" stroke-width="6" stroke-linecap="round" stroke="url(#gradient)"></line></svg>"`;
